finality-aleph/src/validator_network/mod.rs

```diff
@@ -0,0 +1,69 @@
+#![allow(dead_code)]
+use std::fmt::Display;
+
+use aleph_primitives::AuthorityId;
+use codec::Codec;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+mod io;
+#[cfg(test)]
+mod mock;
+
+/// What the data sent using the network has to satisfy.
+pub trait Data: Clone + Codec + Send + Sync + 'static {}
+
+impl<D: Clone + Codec + Send + Sync + 'static> Data for D {}
+
+/// Network represents an interface for opening and closing connections with other Validators,
+/// and sending direct messages between them.
+///
+/// Note on Network reliability and security: it is neither assumed that the sent messages must be
+/// always delivered, nor the established connections must be secure in any way. The Network
+/// implementation might fail to deliver any specific message, so messages have to be resend while
+/// they still should be delivered.
+#[async_trait::async_trait]
+pub trait Network<A: Data, D: Data>: Send {
+    /// Add the peer to the set of connected peers.
+    fn add_connection(&mut self, peer: AuthorityId, addresses: Vec<A>);
+
+    /// Remove the peer from the set of connected peers and close the connection.
+    fn remove_connection(&mut self, peer: AuthorityId);
+
+    /// Send a message to a single peer.
+    /// This function should be implemented in a non-blocking manner.
+    fn send(&self, data: D, recipient: AuthorityId);
+
+    /// Receive a message from the network.
+    async fn next(&mut self) -> Option<D>;
+}
+
+/// A stream that can be split into a sending and receiving part.
+pub trait Splittable: AsyncWrite + AsyncRead + Unpin + Send {
```

Comment:
Do we need this? What's wrong with [split](https://docs.rs/tokio/latest/tokio/io/fn.split.html) since we are just wrapping tokio's types?

Reply:
I am a bit confused by your answer. [split](https://docs.rs/tokio/latest/tokio/io/fn.split.html) is implemented for anything that implements `AsyncWrite + AsyncRead` which is a superset of types required by this trait. Are this `Unpin` and `Send` also necessary - they just make it harder to implement and can be moved to clients of that api (if the need them at all).